src/analysis/souffle/sensitivity_analysis.dl

```diff
@@ -1,90 +1,134 @@
-sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, index) :- 
-      isSqlOperation(operation), 
-      operationHasOperandAtIndex(operation, sql_input, index),
-      isSqlInput(operation, sql_input), 
-      sqlValueSensitivity(sql_input, input_sensitivity_score), 
-      sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, index-1),
-      sensitivity_score >= input_sensitivity_score. 
-
 
-sqlOperationHasSensitivityScoreAtIndex(operation, sensitivity_score, index) :- 
-      isSqlOperation(operation), 
-      operationHasOperandAtIndex(operation, sql_input, index),
-      isSqlInput(value, sql_input), 
-      sqlValueSensitivity(sql_input, sensitivity_score), 
-      sqlOperationHasSensitivityScoreAtIndex(operation, prev_sensitivity_score, index-1), 
-      sensitivity_score >= prev_sensitivity_score. 
+// Inductive cases.  
+// We keep track at the score at each index, and ensure that the sensitivity score at index is only 
+// updated if the input at index has sensitivity score greater than the previous sensitivity score.
+// Otherwise, the prev_index sensitivity score is copied at index. 
+sqlOperationHasSensitivityScoreAtIndex(operation, max(prev_sensitivity_score, input_sensitivity_score), prev_index + 1) :-
+  isSqlOperation(operation),
+  operationHasOperandAtIndex(operation, sql_input, prev_index + 1),
+  isSqlInput(operation, sql_input),
+  sqlValueSensitivity(sql_input, input_sensitivity_score),
+  sqlOperationHasSensitivityScoreAtIndex(operation, prev_sensitivity_score, prev_index). 
 
-// function sensitivity scores per each sql operator as defined in PINQ
+// Function sensitivity scores per each sql operator as defined in PINQ
 // TODO: add other cases 
-.decl sqlFunctionSensitivity(operator: Operator, sensitivity_score: number)
-
-sqlFunctionSensitivity(operator, 1) :- operator = "sql.filter"; 
-      operator = "sql.where"; 
-      operator = "sql.select".
-sqlFunctionSensitivity(operator, 2) :- operator = "sql.group_by". 
+.decl sqlFunctionSensitivity(operator: Operator, sensitivity_score: SensitvityScore)
+
+sqlFunctionSensitivity(operator, 1) :-
+      isAggregationOperation(operation),
+      operationHasOperator(operation, operator).
+sqlFunctionSensitivity(SQL_FILTER_OPERATOR, 1).
+sqlFunctionSensitivity(SQL_WHERE_OPERATOR, 1).
+sqlFunctionSensitivity(SQL_SELECT_OPERATOR, 1). 
+sqlFunctionSensitivity(SQL_COLUMN_ACCESS_OPERATOR, 1). 
+sqlFunctionSensitivity(SQL_OUTPUT_OPERATOR, 1). 
+sqlFunctionSensitivity(SQL_GROUP_BY_OPERATOR, 2).
+
+// The final value sensitivity score, equal to the max of input 
+// multiplied by the function sensitivity.
+.decl sqlValueSensitivity(value: AccessPath, sensitivity_score: SensitvityScore)
 
-// final value sensitivity score, the max of input multiplied by the function sensitivity 
-.decl sqlValueSensitivity(value: AccessPath, sensitivity_score: number)
 sqlValueSensitivity(value, sensitivity_score) :- 
```

Comment:
As a minor point, I think I would find this easier to read if this relation definition comes *before* the definition of `sqlOperationHasSensitivityScoreAtIndex` since `sqlValueSensitivity` is the main one that is consumed by the epsilon part of the analysis.

Reply:
Fixed!